components/resource/scenemanager.cpp

```diff
@@ -244,6 +244,21 @@ namespace Resource
     void SceneManager::setForceShaders(bool force)
     {
         mForceShaders = force;
+
+        if (force)
+        {
+            osg::ref_ptr<Shader::ShaderVisitor> shaderVisitor(createShaderVisitor());
+            shaderVisitor->setDefaults(nullptr);
+            osg::ref_ptr<osg::Node> dummy = new osg::Group;
+            Shader::ShaderVisitor::ShaderRequirements reqs;
```

Comment:
I'm not super keen on using `ShaderRequirements` here from a separation-of-concerns perspective. Previously it was an implementation detail of the shader visitor.

Reply:
Same with the direct call to `createProgram`. As this only runs once and there won't be a measurable performance impact, I'd have created a more convincing dummy stateset and node and used the existing public API instead of making the guts of the shader visible available elsewhere.

Reply:
Using the existing public API with a more thorough dummy setup was my first thought. I realised I would have to separate the dummy setup state from the state created by the ShaderVisitor in this case. We would need at least a dummy node with a dummy stateset and a dummy texture, and a dummy drawable with dummy vertex colors. Using the now public API instead results in less code to maintain, especially if the ShaderVisitor conversion of state into shader requirements will change in the future.

Reply:
Keep in mind we would have to create this dummy setup in more places like groundcover if we expand this PR to other areas. There may not be an easy way to create such dummies in some cases.

Reply:
> We would need at least a dummy node with a dummy stateset and a dummy texture, and a dummy drawable with dummy vertex colors.

I'm pretty sure, but not certain, that more things don't use vertex colours than do, but that's not important as you specify one way or the other with an `osg::Material` either way. The dummy drawable can just be a default-constructed `osg::Geometry` as the shader visitor only needs it to make sure the right overloaded function gets called - it only actually interacts with the geometry in a non-plain-node-compatible way if it needs to create tangents for normal maps, which isn't the default state. Basically, we need next to no code to generate the dummies.

> Using the now public API instead results in less code to maintain, especially if the ShaderVisitor conversion of state into shader requirements will change in the future.

I don't agree at all. We just want to create a reasonably convincing most-common-node, and it's entirely up to the shader visitor which of that state it cares about. Like any other node that gets shaders generated, the shader visitor treats any not-explicitly-set state as left as the default, so there's no need to worry about new state it might care about - the visitor will create a default shader for its default value. With the approach here, however, the scene manager now explicitly *does* need to care about any and all new state that gets tracked as it's got to set it on the shader requirements.

> Keep in mind we would have to create this dummy setup in more places like groundcover if we expand this PR to other areas. There may not be an easy way to create such dummies in some cases.

I think that in every case where we can actually figure out what state combination is most common, and therefore sensible to set a shader for at a local root, we'll have an easier time representing it as a dummy (e.g. type out a copy of what that state is) than manually thinking about what the shader visitor would have done with that state and translating that to shader requirements.